### PR TITLE
feat: add NativeEnum validator

### LINKED ISSUE #48: request: Implement something like zod's nativeEnum
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

Implement something like zod's `nativeEnum`, so we can pass in a TS enum and validate using it (probably with handling transformation of string key -> numeric value too for numeric enums)

### Desired solution

See above

### Alternatives considered

N.A.

### Additional context

_Originally posted by @vladfrangu in https://github.com/sapphiredev/shapeshift/pull/44#discussion_r816033092_

### PR DESCRIPTION
**Pull request by vladfrangu**
Closes #48

### CODE REVIEW

File: tests/validators/nativeEnum.test.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		it('GIVEN a number input for a string enum THEN throws ValidationError', () => {
+			expectError(() => stringPredicate.parse(1), new ValidationError('s.nativeEnum(T)', 'Expected the value to be a string', 1));
+		});
+	});
+
+	describe('number enum', () => {
+		enum NumberEnum {
+			Vladdy,
+			Kyra,
+			Favna
```
Comment (favna) [inline tests/validators/nativeEnum.test.ts:36]: Why am I number 2. I'm way cooler than the Spaniard. 😎😎

**Review by favna — APPROVED** [review]

File: src/validators/NativeEnumValidator.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+		const casted = value as string | number;
+
+		const possibleEnumValue = this.enumMapping.get(casted);
+
+		return typeof possibleEnumValue === 'undefined'
+			? Result.err(
+					new WrongEnumInputError(
+						casted,
+						this.enumKeys.map((key) => [key, this.enumMapping.get(key)!])
```
Comment (kyranet) [inline src/validators/NativeEnumValidator.ts:54]: That's quite an expensive error

File: src/validators/NativeEnumValidator.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		for (const key of this.enumKeys) {
+			const enumValue = enumShape[key] as T[keyof T];
+
+			if (typeof enumValue === 'number') {
+				this.hasNumericElements = true;
+			}
+
+			this.enumMapping.set(key, enumValue);
+			this.enumMapping.set(enumValue, enumValue);
+			this.enumMapping.set(`${enumValue}`, enumValue);
```
Comment (kyranet) [inline src/validators/NativeEnumValidator.ts:29]:
```suggestion
			this.enumMapping.set(key, enumValue);
			this.enumMapping.set(enumValue, enumValue);

			if (typeof enumValue === 'number') {
				this.hasNumericElements = true;
				this.enumMapping.set(`${enumValue}`, enumValue);
			}
```

File: src/validators/BaseValidator.ts [hunk 4]
```diff
@@ -67,7 +68,7 @@ export abstract class BaseValidator<T> {
 		return Reflect.construct(this.constructor, [this.constraints]);
 	}
 
-	protected abstract handle(value: unknown): Result<T, ValidationError | CombinedError | CombinedPropertyError>;
+	protected abstract handle(value: unknown): Result<T, ValidationError | CombinedError | CombinedPropertyError | UnknownEnumValueError>;
```
Comment (kyranet) [inline src/validators/BaseValidator.ts:71]: At this point, we should make this a type so we don't have to update all base validators.

File: src/validators/NativeEnumValidator.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	private readonly enumKeys: string[];
+	private readonly enumMapping = new Map<string | number, T[keyof T]>();
+
+	public constructor(enumShape: T) {
+		super();
+		this.enumShape = enumShape;
+
+		this.enumKeys = Object.keys(enumShape).filter((key) => {
+			return typeof enumShape[enumShape[key]] !== 'number';
+		});
```
Comment (kyranet) [inline src/validators/NativeEnumValidator.ts:18]: ... what is this?

File: src/validators/NativeEnumValidator.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { UnknownEnumValueError } from '../lib/errors/UnknownEnumValueError';
+import { ValidationError } from '../lib/errors/ValidationError';
+import { Result } from '../lib/Result';
+import { BaseValidator } from './imports';
+
+export class NativeEnumValidator<T extends NativeEnumLike> extends BaseValidator<T[keyof T]> {
```
Comment (kyranet) [inline src/validators/NativeEnumValidator.ts:6]: Missing clone (and tests)

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/validators/NativeEnumValidator.ts (hunk 2 above)
Comment (vladfrangu) [reply src/validators/NativeEnumValidator.ts:54]: I guess I could just pass the keys and mapping to the error and only show it on inspect?

File: src/validators/NativeEnumValidator.ts (hunk 5 above)
Comment (vladfrangu) [reply src/validators/NativeEnumValidator.ts:18]: The enum keys

File: tests/validators/nativeEnum.test.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+	describe('valid input but invalid enum value', () => {
+		const predicate = s.nativeEnum({ owo: 42 });
+
+		test.each(['uwu', 69])('GIVEN valid type for input but not part of enum (%p) THEN throws ValidationError', (value) => {
+			expectError(() => predicate.parse(value), new UnknownEnumValueError(value, ['owo'], new Map([['owo', 42]])));
+		});
+	});
+
+	test('GIVEN clone THEN returns similar instance', () => {
```
Comment (imranbarbhuiya) [inline tests/validators/nativeEnum.test.ts:74]: You can use

https://github.com/sapphiredev/shapeshift/blob/793648b4cde1f72c5b735ceebb0c48272179be06/tests/common/macros/comparators.ts#L13

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: tests/validators/nativeEnum.test.ts (hunk 7 above)
Comment (vladfrangu) [reply tests/validators/nativeEnum.test.ts:74]: image

Comment (imranbarbhuiya) [reply tests/validators/nativeEnum.test.ts:74]: Lol sorry 🙊

Btw for consistency you can use `predicate["clone"]()`

Used in ArrayValidator by kyra

Comment (vladfrangu) [reply tests/validators/nativeEnum.test.ts:74]: I copied the code from ObjectValidator which does it this way... Maybe a future pr can go make em all consistent

**Review by favna — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Pull request merged**